Cannot use object attribute on properties

The `object` attribute in the `dbus_proxy` macro is a nice quality-of-life enhancement, but only seems to work on DBus methods. When trying to use it on a DBus property, it produces errors like this:

```
error[E0308]: mismatched types
  --> src/wifi/iwd.rs:60:5
   |
60 |     fn adapter(&self);
   |     ^^^^^^^^^^^^^^^^^- help: consider using a semicolon here: `;`
   |     |
   |     expected `()`, found `Result<_, _>`
   |
   = note: expected unit type `()`
                   found enum `std::result::Result<_, _>`
```

This is from `Device` in the following code:

```rust
#[dbus_proxy(
    interface = "net.connman.iwd.Adapter",
    default_service = "net.connman.iwd",
    assume_defaults = false,
)]
pub trait Adapter {
    #[dbus_proxy(property)]
    fn powered(&self) -> Result<bool>;

    #[dbus_proxy(property)]
    fn set_powered(&mut self, powered: bool) -> Result<()>;

    #[dbus_proxy(property)]
    fn name(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn model(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn vendor(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn supported_modes(&self) -> Result<Vec<String>>;
}

#[dbus_proxy(
    interface = "net.connman.iwd.Device",
    default_service = "net.connman.iwd",
    assume_defaults = false,
)]
pub trait Device {
    #[dbus_proxy(property)]
    fn name(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn address(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn powered(&self) -> Result<bool>;

    #[dbus_proxy(property)]
    fn set_powered(&mut self, powered: bool) -> Result<()>;

    #[dbus_proxy(property, object = "Adapter")]
    fn adapter(&self);

    #[dbus_proxy(property)]
    fn mode(&self) -> Result<String>;

    #[dbus_proxy(property)]
    fn set_mode(&mut self, mode: &str) -> Result<()>;
}
```

Adding `-> Result<OwnedObjectPath>;` fixes the error and allows this code to compile, but it seems the `object` attribute is ignored entirely.

Based on all this, I'm guessing the `object` attribute simply isn't implemented for properties. If possible, it would be a nice enhancement to have.

--- a/zbus/tests/issue/issue_356.rs
+++ b/zbus/tests/issue/issue_356.rs
@@ -1,0 +1,79 @@
+use test_log::test;
+use tracing::instrument;
+use zbus::zvariant::OwnedObjectPath;
+
+/// The "child" interface that will be referenced by the property.
+struct Adapter {
+    name: String,
+}
+
+#[zbus::interface(interface = "org.test.Adapter", proxy(assume_defaults = true))]
+impl Adapter {
+    #[zbus(property)]
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+/// The "parent" interface with a property that returns an ObjectPath to an Adapter.
+struct Device {
+    adapter_path: OwnedObjectPath,
+}
+
+#[zbus::interface(interface = "org.test.Device", proxy(assume_defaults = true))]
+impl Device {
+    /// This property returns an ObjectPath, but the generated proxy should return an AdapterProxy.
+    #[zbus(property, proxy(object = "Adapter"))]
+    fn adapter(&self) -> OwnedObjectPath {
+        self.adapter_path.clone()
+    }
+}
+
+#[test(tokio::test(flavor = "multi_thread"))]
+#[instrument]
+async fn issue_356() {
+    use zbus::connection::Builder;
+
+    let adapter_path: OwnedObjectPath = "/org/test/adapter0".try_into().unwrap();
+    let device_path: OwnedObjectPath = "/org/test/device0".try_into().unwrap();
+
+    // Create the service connection with both interfaces registered
+    let connection = Builder::session()
+        .unwrap()
+        .serve_at(
+            &adapter_path,
+            Adapter {
+                name: "TestAdapter".to_string(),
+            },
+        )
+        .unwrap()
+        .serve_at(
+            &device_path,
+            Device {
+                adapter_path: adapter_path.clone(),
+            },
+        )
+        .unwrap()
+        .name("org.test.Issue356")
+        .unwrap()
+        .build()
+        .await
+        .unwrap();
+
+    // Create a proxy for the Device interface using the auto-generated DeviceProxy
+    let device_proxy = DeviceProxy::builder(&connection)
+        .destination("org.test.Issue356")
+        .unwrap()
+        .path(&device_path)
+        .unwrap()
+        .build()
+        .await
+        .unwrap();
+
+    // Call the adapter() property - this should return an AdapterProxy, not OwnedObjectPath
+    let adapter_proxy = device_proxy.adapter().await.unwrap();
+
+    // Verify we can use the returned proxy to call methods on the Adapter interface
+    let adapter_name = adapter_proxy.name().await.unwrap();
+    assert_eq!(adapter_name, "TestAdapter");
+}

--- a/zbus/tests/issue/mod.rs
+++ b/zbus/tests/issue/mod.rs
@@ -15,6 +15,8 @@ mod issue_81;
 mod issue_279;
 #[cfg(all(unix, feature = "tokio"))]
 mod issue_310;
+#[cfg(all(unix, feature = "tokio"))]
+mod issue_356;
 
 #[cfg(all(unix, feature = "p2p"))]
 mod issue_813;

--- a/zbus_macros/src/iface.rs
+++ b/zbus_macros/src/iface.rs
@@ -1511,7 +1511,7 @@ impl Proxy {
         let member_name = method_info.member_name;
         let mut proxy_method_attrs = quote! { name = #member_name, };
         proxy_method_attrs.extend(match method_info.method_type {
-            MethodType::Signal => quote!(signal),
+            MethodType::Signal => quote!(signal,),
             MethodType::Property(_) => {
                 let emits_changed_signal = properties
                     .get(&member_name)
@@ -1520,7 +1520,7 @@ impl Proxy {
                     .to_string();
                 let emits_changed_signal = quote! { emits_changed_signal = #emits_changed_signal };
 
-                quote! { property(#emits_changed_signal) }
+                quote! { property(#emits_changed_signal), }
             }
             MethodType::Other => quote!(),
         });

--- a/zbus_macros/src/lib.rs
+++ b/zbus_macros/src/lib.rs
@@ -87,8 +87,9 @@ mod utils;
 /// * `allow_interactive_auth` - declare a method call that is allowed to trigger an interactive
 ///   prompt for authorization or confirmation from the receiver.
 ///
-/// * `object` - methods that returns an [`ObjectPath`] can be annotated with the `object` attribute
-///   to specify the proxy object to be constructed from the returned [`ObjectPath`].
+/// * `object` - methods or properties that return an [`ObjectPath`] can be annotated with the
+///   `object` attribute to specify the proxy object to be constructed from the returned
+///   [`ObjectPath`].
 ///
 /// * `async_object` - if the assumptions made by `object` attribute about naming of the
 ///   asynchronous proxy type, don't fit your bill, you can use this to specify its exact name.
@@ -145,6 +146,10 @@ mod utils;
 ///     // `SomeOtherIfaceProxyBlock` would have been assumed and expected. We could also specify
 ///     // the specific name of the asynchronous proxy types, using the `async_object` attribute.
 ///     fn some_method(&self, arg1: &str);
+///
+///     #[zbus(property, object = "SomeOtherIface", blocking_object = "SomeOtherInterfaceBlock")]
+///     // Properties that return an ObjectPath can also use the `object` attribute.
+///     fn related_object(&self);
 /// }
 ///
 /// #[proxy(

--- a/zbus_macros/src/proxy.rs
+++ b/zbus_macros/src/proxy.rs
@@ -246,6 +246,7 @@ pub fn create_proxy(
                     &dbus_member_name,
                     r#_stripped_rust_method_name,
                     m,
+                    &method_attrs,
                     &async_opts,
                     emits_changed_signal,
                 )
@@ -674,6 +675,7 @@ fn gen_proxy_property(
     property_name: &str,
     rust_method_name: &str,
     m: &TraitItemFn,
+    method_attrs: &MethodAttributes,
     async_opts: &AsyncOpts,
     emits_changed_signal: PropertyEmitsChangedSignal,
 ) -> TokenStream {
@@ -699,6 +701,23 @@ fn gen_proxy_property(
             }
         }
     } else {
+        // Check for object attribute to return a proxy instead of OwnedObjectPath
+        let proxy_object = method_attrs.object.as_ref().map(|o| {
+            if *blocking {
+                method_attrs
+                    .blocking_object
+                    .as_ref()
+                    .cloned()
+                    .unwrap_or_else(|| format!("{o}ProxyBlocking"))
+            } else {
+                method_attrs
+                    .async_object
+                    .as_ref()
+                    .cloned()
+                    .unwrap_or_else(|| format!("{o}Proxy"))
+            }
+        });
+
         // This should fail to compile only if the return type is wrong,
         // so use that as the span.
         let body_span = if let ReturnType::Type(_, ty) = &signature.output {
@@ -706,13 +725,36 @@ fn gen_proxy_property(
         } else {
             signature.span()
         };
-        let body = quote_spanned! {body_span =>
-            ::std::result::Result::Ok(self.0.get_property(#property_name)#wait?)
-        };
+
         let ret_type = if let ReturnType::Type(_, ty) = &signature.output {
             Some(ty)
         } else {
             None
+        };
+
+        // Generate the getter body - either returning a proxy object or the raw property value
+        let (body, custom_signature) = if let Some(proxy_path_str) = proxy_object {
+            let proxy_path: Path = parse_str(&proxy_path_str).unwrap();
+            let method_name = Ident::new(rust_method_name, m.sig.ident.span());
+            let inputs = &signature.inputs;
+            let custom_sig = quote! {
+                fn #method_name(#inputs) -> #zbus::Result<#proxy_path<'p>>
+            };
+            let body = quote_spanned! {body_span =>
+                let object_path: #zbus::zvariant::OwnedObjectPath =
+                    self.0.get_property(#property_name)#wait?;
+                #proxy_path::builder(&self.0.connection())
+                    .destination(self.0.destination().to_owned())?
+                    .path(object_path)?
+                    .build()
+                    #wait
+            };
+            (body, Some(custom_sig))
+        } else {
+            let body = quote_spanned! {body_span =>
+                ::std::result::Result::Ok(self.0.get_property(#property_name)#wait?)
+            };
+            (body, None)
         };
 
         let (proxy_name, prop_stream) = if *blocking {
@@ -771,16 +813,28 @@ fn gen_proxy_property(
             PropertyEmitsChangedSignal::False => quote! {},
         };
 
+        // When object attribute is used, we use a custom signature and skip cached/receive methods
+        // since those don't make sense for proxy objects
+        let (sig, extra_methods) = if let Some(sig) = custom_signature {
+            (sig, quote! {})
+        } else {
+            (
+                quote! { #signature },
+                quote! {
+                    #cached_getter_method
+                    #receive_method
+                },
+            )
+        };
+
         quote! {
             #(#other_attrs)*
             #[allow(clippy::needless_question_mark)]
-            pub #usage #signature {
+            pub #usage #sig {
                 #body
             }
 
-            #cached_getter_method
-
-            #receive_method
+            #extra_methods
         }
     }
 }

